WIP: Data Access Pattern in 0.5

Evizero opened the pull request:
WIP implementation for #13 
---

I have updated `DataSubset` and its tests to 0.5 so far.
- [x] still needs updated documentation
- [x] still need to test `datasubset`

Next I'll port Tom's new verbs (in this PR)
- [x] `eachobs` and `shuffled` implementation and tests
- [x] `eachbatch` documentation and tests
- [x] update documentation to Tuple change
- [x] `DataIterator` documentation
- [x] `splitobs` and `batches` implementation and tests
- [x] `viewobs` implementation and tests
- [x] Split `DataIterator` into `BatchIterator` and `ObsIterator`
- [ ] `filterobs` implementation and tests
- [ ] `infinite_batches` and `infinite_obs`
- [x] `KFolds`, `kfolds`, `leaveout` implementation and tests

Also I shall try to update the package documentation a bit and try avoid breaking changes if I can
- [ ] docstrings for `getobs` and `nobs`
- [ ] Try to deprecate old API without breaking code
- [ ] Update Readme to reflect new changes
- [x] Nicer show methods for custom types

tbreloff commented on REQUIRE:
```diff
@@ -1,2 +1,3 @@
-julia 0.4
+julia 0.5
+LearnBase
```
need to add a dependency on Iterators?

Evizero replied:
yes

tbreloff commented on src/MLDataUtils.jl:
```diff
@@ -1,9 +1,16 @@
+using LearnBase
 using StatsBase
 
+import Iterators: repeatedly, repeated
```
it would be nice to re-export these

Evizero replied:
👍

tbreloff commented on src/accesspattern/arrays.jl:
```diff
@@ -0,0 +1,24 @@
+
+# apply a view to the last dimension
+@generated function getobs(A::AbstractArray, idx)
+    T, N = A.parameters # TODO: This is error prone. the subtype may have different ordering or parameters. See SparseMatrixCSC
```
good point

tbreloff commented on src/accesspattern/arrays.jl:
```diff
@@ -0,0 +1,24 @@
+typealias NativeArray Union{Array,SubArray}
```
I'm a little hesitant here... what about the many other array types, like PermutedDimsArray?

Evizero replied:
Yea, I agree it is sub optimal. And I dislike the name, although it is just internal. I am unsure how to best include all fitting types, but I rather be too restrictive than too loose (example Sparse Arrays). `DataSubset` shouldn't slow anything down really

tbreloff commented on src/accesspattern/datasubset.jl:
```diff
@@ -0,0 +1,156 @@
+        if T <: Tuple
+            length(unique(map(_->nobs(_), data))) == 1 || throw(DimensionMismatch("all parameters must have the same number of observations"))
+        end
+        1 <= minimum(indices) || throw(BoundsError(data, indices))
```
it's possible this isn't even valid with all the funky arrays @timholy has coming... i think we should leave it out and let it error if it's out of bounds on access.

Evizero replied:
I don't think `OffsetArrays` and alike make much sense in our use-cases, if that is what you mean. Otherwise we have bigger problems all over the place I think. Starting with computing batch ranges etc.

Evizero replied:
I suppose we could make an effort into generalizing to arbitrarily indexed arrays. But let's do that in a different PR

tbreloff commented on src/accesspattern/tuples.jl:
```diff
@@ -0,0 +1,51 @@
+@generated function datasubset(A::Tuple, indices)
+    if all(map(T -> T<:NativeArray, A.types))
+        # Don't box in DataSubset
+        expr = :(length(unique(map(a->nobs(a), A))) == 1 || throw(DimensionMismatch("all parameters must have the same number of observations")))
```
Could we maybe replace these calls with a non-allocating version?

``` julia
function check_nobs(A::Tuple)
    n1 = length(A[1])
    for i=2:length(A)
        if length(A[2]) != n1
            throw(DimensionMismatch("all parameters must have the same number of observations"))
        end
    end
end
```

Evizero replied:
yeah, surely wouldn't hurt

tbreloff commented:
@Evizero overall this is coming along really nicely... thanks for putting in the time

Evizero commented:
I had to introduce a immutable type `DataIterator` to make `eachobs` work. This is a consequence of not boxing Tuple into `DataSubset`. The alternative would be to call batches with batch-size 1, which seems like a waste, since it allocates an array of N DataSubsets

tbreloff commented:
I saw... certainly better than using DataSubsets!

Evizero commented:
I think I could twist this one type `DataIterator` in a way that it could also serve as a batch stream. Then we could offer `eachbatch` as well, which would be a true iterator, while `batches` is done instantly and remains as it is now

Evizero commented:
shoo coveralls. come back when I am done here. You make me look bad!

![shoo](http://rs71.pbsrc.com/albums/i140/shortpenguin/Gifs/Movies/Pirates%20of%20the%20Caribbean/tumblr_lx1jycL0tG1qeqseto1_250_zpsbc2ca5de.gif~c200)

Evizero commented:
Ok, so the stuff I have implemented so far works pretty well and type stable. `KFolds` might be a little tricky to port, but other than that it should be smooth sailing now.

tbreloff commented:
Nice!

On Tuesday, October 18, 2016, Christof Stocker notifications@github.com
wrote:

> Ok, so the stuff I have implemented so far works pretty well and type
> stable. KFolds might be a little tricky to port, but other than that it
> should be smooth sailing now.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> https://github.com/JuliaML/MLDataUtils.jl/pull/14#issuecomment-254673866,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AA492pkBlEt8KHkHhkRWxAQmLdAvd88Bks5q1V42gaJpZM4KZHBf
> .

tbreloff commented:
Allowing both `getobs` and `viewobs` is a nice idea (as long as viewobs is the default!)

Evizero commented:
I changed how `getobs` works a bit. 
- Calling `getobs(A::Any)` will return `A` itself, but calling `getobs(A::SubArray)` will return `copy(A)` (and `getobs(s::DataSubset)` in `getobs(s.data, s.indcies)`). 
  This way the user can just call `getobs(..)` in the inner most loop before using the data/batch in order to benefit from cache locality etc.
- To offer an alternative I also introduced `viewobs` which is just a different name for `datasubset`

Evizero commented:
> as long as viewobs is the default!

They are not really used internally ~~expect from the function `eachobs`, which sounds like one would expect it to `getobs` every index~~ Edit: no longer. So it is just user-facing code. All the functions such as `batches` and `shuffled` remain lazy as ever and call `datasubset` directly (which to be fair is the same as `viewobs` atm, but that may change).

tbreloff commented:
> remain lazy as ever and call datasubset directly

That's what I mean... I want to make sure everything is lazy unless you explicitly ask for copies through collect/getobs.

tbreloff commented:
I checked out refactor0.5, and commented out the iterators in StochasticOptimization.  I'll try to get the surrounding ecosystem up to speed and tested.

Evizero commented:
nice! There is still some new stuff missing and I will need to work on a different project for a few hours soon. But I will resume this PR later today.

tbreloff commented:
No worries... I think I'm the only one using this stuff for any real projects. (I mean... I **hope** I'm the only one :smile: )

tbreloff commented:
Question... can DataIterator and DataSubset at least subtype from a common abstract?  AbstractSubset?  They seem to have very similar behavior... maybe I don't quite understand their distinction yet

tbreloff commented on src/accesspattern/dataiterator.jl:
```diff
@@ -0,0 +1,175 @@
+Base.start(iter::DataIterator) = iter.start
+Base.done(iter::DataIterator, idx) = maximum(idx) > iter.count * length(iter.start)
+Base.next(iter::DataIterator, idx::UnitRange) = (datasubset(iter.data, idx), idx + length(iter.start))
+Base.next(iter::DataIterator, idx::Int) = (getobs(iter.data, idx), idx + length(iter.start))
```
Should this be either viewobs or datasubset, now that getobs copies?

Evizero replied:
That is what I meant with `eachobs` is special and sounds like it should return `getobs` for each index. I don't have a strong opinion on this though. It does seem a bit weird to return a view for a single observation. what would one be waiting for?

tbreloff replied:
If a "single observation" is a giant 3D volume of spatial data, you may not want to copy it.

tbreloff replied:
Should we have 2 different methods?  One that copies and one that doesn't?  And add that as a type parameter?

``` julia
type DataIterator{T,I,COPY}
  ...
end
...
Base.next{T,I}(itr::DataIterator{T,I,true}, i) = (getobs(itr,i), i+1)
Base.next{T,I}(itr::DataIterator{T,I,false}, i) = (viewobs(itr,i), i+1)
```

Evizero replied:
I suppose there is no hurt in leaving the copying as opt-in with `getobs`. Would maybe be better anyway to be consistent among all functions in that the return `views`.

I will change to `datasubset`

tbreloff commented on src/accesspattern/datasubset.jl:
```diff
@@ -0,0 +1,185 @@
+
+Base.start(::DataSubset) = 1
+Base.done(subset::DataSubset, idx) = idx > length(subset.indices)
+Base.next(subset::DataSubset, idx) = (subset[idx], idx + 1)
```
Should this be viewobs/datasubset?

Evizero replied:
see my comment above

Evizero commented:
Well to be fair, `DataIterator` doesn't need `getobs` and `nobs`, and `DataSubset` doesn't need to support the iterator pattern. These thing are only implemented because they can be.
`DataSubset` alone is not enough since there is no reliable way to implement `eachobs` then for tuples, or `eachbatch` for that matter.

tbreloff commented:
> DataSubset alone is not enough since there is no reliable way to implement eachobs then for tuples, or eachbatch for that matter.

I'm still confused here... I thought my implementation in StochasticOptimization handled any type just fine.

Evizero commented:
yea, but you boxed `Tuple` into a `DataSubset`, while this doesn't. This is a consequence of allowing types to offer their own views if they have such, which is the reason that if one works with `Array` variables, that one never has to deal with `DataSubset`

tbreloff commented:
This does a copy, which I was really hoping to avoid:

```
julia> using MLDataUtils
INFO: Recompiling stale cache file /home/tom/.julia/lib/v0.5/MLDataUtils.ji for module MLDataUtils.

julia> x,y = rand(10,2), rand(10)
(
[0.523045 0.975809; 0.963106 0.955312; … ; 0.215673 0.988101; 0.555086 0.376569],

[0.582727,0.160317,0.983288,0.0741725,0.789932,0.841186,0.119106,0.534921,0.0328525,0.0659002])

julia> for o in eachobs(x)
           @show typeof(o), o
       end
(typeof(o),o) = (Array{Float64,1},[0.523045,0.963106,0.0167039,0.928771,0.726239,0.656165,0.976564,0.350933,0.215673,0.555086])
(typeof(o),o) = (Array{Float64,1},[0.975809,0.955312,0.85547,0.578021,0.959107,0.769584,0.114884,0.31311,0.988101,0.376569])
```

Evizero commented:
this is exactly our discussion above, which I will now change to not copy

tbreloff commented:
> I had to introduce a immutable type DataIterator to make eachobs work. This is a consequence of not boxing Tuple into DataSubset.

Reading back through the conversation, I think this is what I don't yet understand.  My implementation worked well (I thought) by boxing the tuple inside the DataSubset... I don't see why a tuple of DataSubsets would be more performant.  What functionality didn't work?

Evizero commented:
I would need to either box everything into a DataSubset again, or overload functions for tuples, which would interfere with outside code when MLDataUtils is loaded

tbreloff commented:
Are you talking specifically about the case of `eachbatch`-style iteration?  You mean that each batch would need to be re-wrapped with a DataSubset, but you want to return a tuple of views?

Evizero commented:
I am saying that I would like the following

``` julia
X, Y = load_iris()

# tuple in results tuple out
tup = datasubset((X,Y), 1:10)
@assert typeof(tup) <: Tuple

x_sub, y_sub = datasubset((X,Y), 1:10)
@assert typeof(x_sub) <: SubArray
@assert typeof(y_sub) <: SubArray

train, test = splitobs(X,Y)
# train and test are Tuple of SubArrays 
```

Now the problem arises for this kind of use:

``` julia
for (x,y) in eachobs(X,Y)
    # ...
end
```

if `eachobs` calls `datasubset`, it would return a `Tuple`.
The code above would then call `start`, `next` etc for the Tuple. I would like to avoid overloading the iterator functions for `Tuple`, hence `eachobs` returns a `DataIterator`

tbreloff commented:
Ok I'm playing around with it and warming up to the idea of always returning tuples/views.  I guess I'm wondering if we can merge the DataIterator and DataSubset types.  Do they really need to be separate?

Evizero commented:
If I find a way to remove one cleanly, I will

tbreloff commented:
So the first problem I'm up against... in StochasticOptimization I dispatch the `search_direction` method on whether I have a single observation or an AbstractSubset (which implies it's a batch).  The problem is, I have no idea what a "single observation" looks like... only that it's not an AbstractSubset.  A single observation might be a Float64, or it might be a tuple of 3D arrays.  So if my loop was essentially:

``` julia
for batch in eachbatch(data, size=10)
    ...
    sd = search_direction(batch)
    ...
end
```

then I have no idea whether I'm processing one observation or one batch purely on the types.

tbreloff commented:
What if `batches`/`splitobs` returned a DataIterator as well, and we just expand the parameters a bit to include the DataSubset functionality at the same time?  The goal would be one type that represents "something to be iterated over", and a type parameter from which we could say whether we are going to get observations or batches.  I'll play with that idea and see if I can come up with something.

Evizero commented:
Then we have the same situation again with the tuples. mhm. I'll think about the problem a bit after work.

Generally I would like to offer a user as little complexity as possible and make other things opt-in.

tbreloff commented:
Well if I can dispatch at the MetaLearner call on what type of iterator is passed in (i.e. iterating over observations or batches), then I can somehow pass that information through myself.  It's certainly more clunky and the implementation leaks, but it could work.  And I agree that most people would prefer the behavior you're going for.  I think I have an idea... I'll implement it and push it up as a separate branch so you can look.

Evizero commented:
What if we offer a typed API that starts with an upper case? like `Eachobs` and `Eachbatch`, which return specials types to dispatch on, but internally use the lower case methods to perform their duties

tbreloff commented:
> What if we offer a typed API start starts with an upper case? like Eachobs and Eachbatch, which return specials types to dispatch on, but internally use the lower case methods to perform their duties

Then we'd be forcing the users to always use an uppercase version when dispatch is needed?  Seems confusing and fragile.  I'm working on fixing this and merging DataSubset into DataIterator... give me an hour or so.

Evizero commented:
Like `Eachobs` iterates over `Observation`, and `Eachbatch` over `Batch`, maybe `Batches` returns a `Vector{Batch}`, etc

tbreloff commented:
I think this is working out well.  The changes:
- call viewobs internally to get unwrapped views... datasubset now **always** returns a DataIterator
- remove the datasubset.jl file
- the `S` parameter of DataIterator can be a vector of Int (old DataSubset) or a vector of Int-vectors (old DataSubsets)
- dispatch iteration on the S parameter so that we get the correct behavior of eachobs/eachbatch/DataSubset/batches/splitobs
- same dispatch for `Base.show` to more clearly show what type of iterator we have
- datasubset/batches/splitobs all construct an appropriate DataIterator
- added an `isbatches` method which tells us if we're getting an observation or a batch at each iteration

Here's a demo:

``` julia
julia> using MLDataUtils; n=5;  x, y = rand(10,n),rand(n);

julia> datasubset(x)
DataIterator{Array{Float64,2}}: 5 observations

julia> eachobs(x)
DataIterator{Array{Float64,2}}: 5 observations

julia> datasubset(x, [1,2])
DataIterator{Array{Float64,2}}: 2 observations

julia> datasubset(x, 1:2)
DataIterator{Array{Float64,2}}: 2 batches with 2 obs each

julia> eachbatch(x)
DataIterator{Array{Float64,2}}: 5 batches with 1 obs each

julia> eachbatch(x,size=2)
INFO: The specified values for size and/or count will result in 1 unused data points
DataIterator{Array{Float64,2}}: 2 batches with 2 obs each

julia> typeof(shuffled(x))
SubArray{Float64,2,Array{Float64,2},Tuple{Colon,Array{Int64,1}},false}

julia> batches(x)
DataIterator{Array{Float64,2}}: 5 batches with indices: UnitRange{Int64}[1:1,2:2,3:3,4:4,5:5]

julia> batches(x, size=2)
INFO: The specified values for size and/or count will result in 1 unused data points
DataIterator{Array{Float64,2}}: 2 batches with indices: UnitRange{Int64}[1:2,3:4]

julia> MLDataUtils.isbatches(ans)
true

julia> eachbatch(x,size=2)
INFO: The specified values for size and/or count will result in 1 unused data points
DataIterator{Array{Float64,2}}: 2 batches with 2 obs each

julia> MLDataUtils.isbatches(ans)
true

julia> datasubset(x, [1,2])
DataIterator{Array{Float64,2}}: 2 observations

julia> MLDataUtils.isbatches(ans)
false

julia> splitobs(x, at=(0.2,0.3,0.2))
DataIterator{Array{Float64,2}}: 4 batches with indices: UnitRange{Int64}[1:1,2:3,4:4,5:5]

julia> MLDataUtils.isbatches(ans)
true

julia> for b in splitobs(x, at=0.3)
           @show typeof(b) b
       end
typeof(b) = SubArray{Float64,2,Array{Float64,2},Tuple{Colon,UnitRange{Int64}},true}
b = [0.84198 0.512395; 0.447355 0.324533; 0.671367 0.305329; 0.606857 0.418458; 0.727636 0.738322; 0.836541 0.302326; 0.040053 0.254213; 0.572415 0.874371; 0.884836 0.00895098; 0.652987 0.658197]
typeof(b) = SubArray{Float64,2,Array{Float64,2},Tuple{Colon,UnitRange{Int64}},true}
b = [0.849606 0.0897501 0.256851; 0.192671 0.466011 0.737302; 0.121008 0.00902418 0.125703; 0.18939 0.0111741 0.397602; 0.295394 0.00677543 0.854273; 0.946626 0.967494 0.0430847; 0.213562 0.760838 0.468581; 0.00534368 0.40582 0.650375; 0.50318 0.987332 0.0669072; 0.79452 0.0297965 0.73686]
```

tbreloff commented:
f68fd32

Evizero commented:
> julia> splitobs(x, at=(0.2,0.3,0.2))
> DataIterator{Array{Float64,2}}: 4 batches with indices: UnitRange{Int64}[1:1,2:3,4:4,5:5]

This is exactly what I want to avoid though

tbreloff commented:
> This is exactly what I want to avoid though

Can you explain why?  This seems to be an ideal result:

``` julia
julia> test,train = splitobs(x, at=0.6)
DataIterator{Array{Float64,2}}: 2 batches with indices: UnitRange{Int64}[1:3,4:5]

julia> typeof(test)
SubArray{Float64,2,Array{Float64,2},Tuple{Colon,UnitRange{Int64}},true}

julia> typeof(train)
SubArray{Float64,2,Array{Float64,2},Tuple{Colon,UnitRange{Int64}},true}
```

Evizero commented:
I'd rather not have everything be an iterator. Especially not when the user decides to do something like `sub = viewobs(X, 1:4)` if X is an Array

tbreloff commented:
``` julia
julia> viewobs(x) |> typeof
Array{Float64,2}

julia> viewobs(x, 1:4) |> typeof
SubArray{Float64,2,Array{Float64,2},Tuple{Colon,UnitRange{Int64}},true}

```

tbreloff commented:
- getobs/viewobs return arrays/tuples
- eachobs/datasubset return DataIterator over observations
- eachbatch/batches/splitobs return DataIterator over batches

Evizero commented:
Ok, I obviously need to take a look at the actual code and not just the few examples you posted.
My conference call will end soon and I shall give it an honest joyride. Sorry for half-informed pre-emptive judgement.

tbreloff commented:
And I think we could take this one step further and do:

``` julia
abstract DataIterator
type ObsIterator <: DataIterator ... end
type BatchIterator <: DataIterator ... end
```

and keep most of the same code except some of the bonus methods like `infinite_obs` and `infinite_batches` would have cleaner implementations.

tbreloff commented:
I added another param, and an abstraction, (6593e12) and now the infinite methods are working:

```
julia> for (i,b) in enumerate(infinite_batches(x,y,size=2))
           @show i nobs(b) typeof(b)
           if i>5
               break
           end
       end
i = 1
nobs(b) = 2
typeof(b) = Tuple{SubArray{Float64,2,Array{Float64,2},Tuple{Colon,Array{Int64,1}},false},SubArray{Float64,1,Array{Float64,1},Tuple{Array{Int64,1}},false}}
i = 2
nobs(b) = 2
typeof(b) = Tuple{SubArray{Float64,2,Array{Float64,2},Tuple{Colon,Array{Int64,1}},false},SubArray{Float64,1,Array{Float64,1},Tuple{Array{Int64,1}},false}}
i = 3
nobs(b) = 2
typeof(b) = Tuple{SubArray{Float64,2,Array{Float64,2},Tuple{Colon,Array{Int64,1}},false},SubArray{Float64,1,Array{Float64,1},Tuple{Array{Int64,1}},false}}
i = 4
nobs(b) = 2
typeof(b) = Tuple{SubArray{Float64,2,Array{Float64,2},Tuple{Colon,Array{Int64,1}},false},SubArray{Float64,1,Array{Float64,1},Tuple{Array{Int64,1}},false}}
i = 5
nobs(b) = 2
typeof(b) = Tuple{SubArray{Float64,2,Array{Float64,2},Tuple{Colon,Array{Int64,1}},false},SubArray{Float64,1,Array{Float64,1},Tuple{Array{Int64,1}},false}}
i = 6
nobs(b) = 2
typeof(b) = Tuple{SubArray{Float64,2,Array{Float64,2},Tuple{Colon,Array{Int64,1}},false},SubArray{Float64,1,Array{Float64,1},Tuple{Array{Int64,1}},false}}


julia> for (i,b) in enumerate(infinite_obs(x,y))
           @show i typeof(b)
           if i>5
               break
           end
       end
i = 1
typeof(b) = Tuple{SubArray{Float64,1,Array{Float64,2},Tuple{Colon,Int64},true},Float64}
i = 2
typeof(b) = Tuple{SubArray{Float64,1,Array{Float64,2},Tuple{Colon,Int64},true},Float64}
i = 3
typeof(b) = Tuple{SubArray{Float64,1,Array{Float64,2},Tuple{Colon,Int64},true},Float64}
i = 4
typeof(b) = Tuple{SubArray{Float64,1,Array{Float64,2},Tuple{Colon,Int64},true},Float64}
i = 5
typeof(b) = Tuple{SubArray{Float64,1,Array{Float64,2},Tuple{Colon,Int64},true},Float64}
i = 6
typeof(b) = Tuple{SubArray{Float64,1,Array{Float64,2},Tuple{Colon,Int64},true},Float64}
```

tbreloff commented:
With the `tb` branch:

```
INFO: StochasticOptimization tests passed
```

tbreloff commented:
Also, my Reinforce.jl code works as well.  So I'm going to go back to working on that, and we can discuss how to move forward with MLDataUtils later when you're ready.

Evizero commented:
A few thoughts
- `splitobs` and `batches` returning an iterator just feels wrong (what's the difference to `eachbatch` then?), and I think I will make them behave as they are now, returning a Vector.
- `viewobs(my_sparse_array,  1:5)` (which atm throws an error) returning something called a `DataIterator` also doesn't seem right. That could be fixed by changing the name of the Type somewhat.
- what would `eachbatch(datasubset(my_sparse_array))` do? right now it nests `DataIterator` and querying it throws an error. I suppose that could be fixed though.

The more I play with it and think about it the more I want a special "view"/"subset" type that is only concerned with representing a lazy subset of the data, nothing more.
It doesn't really reduce complexity to have one type that behaves like multiple types either, it just makes things more convoluted.

I would go the other way even, introducing a 3rd type by splitting the `DataIterator` that I have up into a "batch" and "obs" version.

Concerning your dispatch problem. If a user is allowed to call your code with plain arrays that can either be batches or single observations, then you need to be able to handle native types anyways. If this sub-setting is done within your code then you have the control to use `eachbatch` and `eachobs` for the dispatchable types.

I really really don't want **everything** to be a true iterator. It feels off and non-julian to me. Using Iterators should be conscious decision I think.

tbreloff commented:
How do you propose implementing the "infinite" iterators?

Evizero commented:
pretty much the same as you

tbreloff commented:
It would be nice if:

```
infinite_obs --> ObsIterator
infinite_batches --> BatchIterator
```

Evizero commented:
I agree. That thought crossed my mind as well

Evizero commented:
The problem is in the detail though. I could make a flag for the `BatchIterator`to be infinite, but for consistency it should then loop over the batches in their true order. The inifite one you have in mind does random subsamples every time. It is conceptually different.

tbreloff commented:
Might need to add an extra abstraction for this to work without complicating the other types:

``` julia
abstract DataIterator
  abstract AbstractObsIterator <: DataIterator
    type ObsIterator <: AbstractObsIterator
    type InfObsIterator <: AbstractObsIterator
  # same for batches
```

Evizero commented:
On that note, can you think of a better name than `ObsIterator` ? I like the type, but the name could be better

Evizero commented:
What if the types reflect the function?

``` julia
abstract DataIterator

abstract BatchIterator <: DataIterator
immutable Eachbatch <: BatchIterator end
immutable Infbatches <: BatchIterator end

abstract ObsIterator <: DataIterator
immutable Eachobs <: ObsIterator end
immutable Infobs <: Obsiterator end
```

tbreloff commented:
> What if the types reflect the function?

Yes I like that.  Nitpick... I'd prefer camelcase: EachObs.

I'm wondering if DataSubset can be:

``` julia
type SubsetObs <: ObsIterator
```

I'm still not convinced that `shuffled` and similar should return `viewobs` and not `datasubset`.

tbreloff commented:
It just really bothers me that DataSubset is a total separate type...

Evizero commented:
> I'm wondering if DataSubset can be:

I completely removed the ability to iterate over a `DataSubset`. you don't iterate over a `SubArray` either, so it doesn't quite add up that it might work when using custom types but not for plain arrays. `DataSubset` is now just a lazy view and serves as a fall back for types that can't offer such.

tbreloff commented:
Sure you can iterate over a subarray just like any other array:

``` julia
julia> x = rand(10)
10-element Array{Float64,1}:
 0.879267 
 0.7712   
 0.60451  
 0.998378 
 0.674976 
 0.0372644
 0.150353 
 0.460558 
 0.0960781
 0.490523 

julia> y = view(x,2:4)
3-element SubArray{Float64,1,Array{Float64,1},Tuple{UnitRange{Int64}},true}:
 0.7712  
 0.60451 
 0.998378

julia> for yi in y
       @show yi
       end
yi = 0.7711995499255015
yi = 0.604509881730505
yi = 0.998377795508409
```

Evizero commented:
my point is this:

``` julia
julia> for x in rand(2,5)
       @show x
       end
x = 0.48582956183939885
x = 0.8691764010130829
x = 0.15545875870878745
x = 0.09107139756381688
x = 0.09622235915747268
x = 0.7638308964988003
x = 0.5252940680537728
x = 0.4871337374237992
x = 0.7661749720385271
x = 0.18993652421902274
```

Evizero commented:
code that expects that the given `data` iterates by default over its "observations", will not work for plain arrays. We should enforce `eachobs(data)` for such cases

tbreloff commented:
I don't understand your point...

To be clear, I view data iterators/subsets as roughly `AbstractVector{Observation}`... regardless of what the Observation type is.  So if you pass a 3D array to a ObsIterator, each Observation will be a 2D array.

Evizero commented:
> To be clear, I view data iterators/subsets as roughly AbstractVector{Observation}

I disagree for subsets. For some type I may not know what it represents, just that it is a subset of it. A SubArray isn't a vector of observations either. If I subset some data I expect something that conceptually represents the subset, and not an iterator into the observations of it. If I subset some data I am not asking about an iterator into each observation of that, I am asking about a subset placeholder that at some point down the line some code will know how to deal with what it represents. 

> So if you pass a 3D array to a ObsIterator, each Observation will be a 2D array.

Yes but then you are asking for an iterator into each observation

tbreloff commented:
What purpose do you see for subsets if you won't iterate through them (or otherwise access them like a vector of observations) in some way?   Why are they there?  I think I'm missing your perspective on when you would ever use a subset.

Evizero commented:
One use-case I have in mind is for Augmentor. I will have a type `DirImageSource` which represents a number of images in some directory and it's subdirectories. As an example: I would like to split it into a training and testset, and then maybe further subset 10 random images from the trainingset and then load that subset with `getobs` to plot them as examples images in the training test.

Evizero commented:
or maybe, to move a bit further away from "you could use eachobs for the last step", I want to compute the mean-image from that last subset

Evizero commented:
That doesn't mean that any other code will know how to deal with what my special type returns when `getobs` is called, by the way, there is no such contract. The patterns should just be an easy tool for people to solve their own problems more conveniently, and all they have to do to use these goodies is implement `getobs` and `nobs`

tbreloff commented:
> One use-case I have in mind is for Augmentor.

I'm clearly missing something, because that example seems to perfectly fit the model:

``` julia
train, test = splitobs(data) # splitobs returns a length-2 batch iterator
subbatch = datasubset(train, rand(1:nobs(train), 10)  # could add a randbatch method to do this?
mean(subbatch)
```

which by default would call https://github.com/JuliaLang/julia/blob/master/base/statistics.jl#L10 which depends on an iteration inteface and `+`/`/` being defined for your image types.  If you don't have the iteration, then you don't get all this machinery for free.

Evizero commented:
No, I don't want to work with the subset abstraction, I want to work with the data that is underneath, whatever it is. Otherwise I can always wrap into `eachobs` to get these goodies

tbreloff commented:
> No, I don't want to work with the subset abstraction, I want to work with the data that is underneath, 

Can you go into more detail on your Augmentor example, and why you think the subset abstraction will make things more complicated?  What (precisely) are the types at each step of your example, and why?

Evizero commented:
It's a conceptual thing. `map` doesn't return an iterator either. `view` doesn't return an iterator. They could, but they don't

Evizero commented:
Not the perfect examples, I grant you though.

tbreloff commented:
> map doesn't return an iterator either. view doesn't return an iterator.

Ummm... yes they do:

``` julia
julia> x = rand(5)
5-element Array{Float64,1}:
 0.38471 
 0.159309
 0.999277
 0.996982
 0.109125

julia> for xi in map(sin,x)
           @show xi
       end
xi = 0.3752900649518715
xi = 0.15863594848736212
xi = 0.8410802667167983
xi = 0.8398366067265394
xi = 0.10890871498222936

julia> for xi in view(x,2:3)
           @show xi
       end
xi = 0.15930895331619754
xi = 0.9992772595118062
```

Evizero commented:
with that logic so does my implementation of `batches`, and `splitobs`, which return a Vector

Evizero commented:
By the way I am not arguing that `DataSubset` must not implement the iterator interface. I removed it to force the use of `eachobs` because otherwise code might easily be brittle when called with `SubArrays` that are suddenly expected to also iterate of observations

tbreloff commented:
More and more I'm wanting everything to return a DataIterator, for consistency.  Now that we (might) have distinct ObsIterator and BatchIterator abstractions, I think the api will be much less confusing if a "Batch" or "DataSubset" is really just an ObsIterator:

```
shuffled(data) --> ObsIterator
eachobs(data) --> ObsIterator
randbatch(data, size=10) --> ObsIterator
bs = batches(data) --> BatchIterator
for batch in bs --> batch is ObsIterator
train,test = splitobs(data) --> train/test are ObsIterators
infbat = infinite_batches(train)  --> BatchIterator
for minibatch in infbat --> minibatch is ObsIterator
```

Now, you don't want this... I get it.  But it **greatly** simplifies the confusion that a user might have.  "I call this function and I get one of 2 types of lazy iterators... ok got it!"

The part you really dislike (I think) is that you have to "extract" your data into a non-MLDataUtils container to do something useful with it.  Well, I don't think that's necessarily true.  And we could make this one step easier if `DataIterator <: AbstractVector`.  If DataIterator really was a vector, and you really do have a `AbstractVector{AugmentedImage}`, then maybe you wouldn't be so desperate to avoid the DataIterator type?

Evizero commented:
> shuffled(data) --> ObsIterator

No reason whatsover to do that for an array which has a perfectly good `SubArray`

Evizero commented:
One thing I dislike is imposing unnecessary Types on potential users that might not need it. Especially not those working with arrays.

Evizero commented:
I'd like to keep the user experience as native as possible. Not the developer experience. One might never encounter DataSubset if one is working with arrays, or one may never encounter the iterators when working with `splitobs` and `batches` which return plain vectors which is a typical thing julia functions do

tbreloff commented:
> One thing I dislike is imposing unnecessary Types on potential users that might not need it. Especially not those working with arrays.

But conceptually, anyone using this API is stating implicitly "Look, this is a 3D Array, but really it's not, it's a vector of 2D arrays, and I want ways to treat this 3D array like a vector of 2D arrays without copying the data around."

So giving them back an `AbstractVector{Matrix}` is exactly what they'd be hoping for in most cases.

Evizero commented:
Who knows, maybe there will be TableViews as well, which we could then support and then those working with DataFrames would also never see unfamiliar types

tbreloff commented:
> I'd like to keep the user experience as native as possible. Not the developer experience. 

Me too... which is why I'm pushing for this.  I think doing developer magic so that you can return SubArrays will make everything more confusing for newcomers.

Evizero commented:
> Look, this is a 3D Array, but really it's not, it's a vector of 2D arrays

I disagree there. If I have a typical image array of let's say size (64,64,3,100)  and i want a subset of 20 out of that, then I don't think of that as a vector of 20 distinct (64,64,3) arrays, I think of it as one (64,64,3,20) array

Evizero commented:
> I think doing developer magic so that you can return SubArrays will make everything more confusing for newcomers

I see it the other way around, it returns familiar types by default. Probably most users working with images or numeric data work with plain arrays. It **can** support special user types if a user wants to opt into that ecosystem. For such it has a fallback-subset

tbreloff commented:
You really think returning different type structures depending on the specifics of the input data as being **more** intuitive?  I think a user will be forever confused as to what they'll get back from a method.

Evizero commented:
> returning different type structures depending on the specifics of the input data

sounds like Julia to me

Evizero commented:
Also, it is not that convoluted. The subset returns a view, either native for the type or a fallback. Pretty consistent I think

tbreloff commented:
> The subset returns a view

Except that it doesn't... for the case of Array, it returns a `view(data, :, :, 3:7)`, **not** a `view(data, 3:7)` which is what a "data subset" is.  This distinction seems natural to you, but I'm pretty confident it's going to be really confusing for your average data scientist that's not also a serious Julia hacker.

Evizero commented:
It does enforce the convention that the last dimension denotes the observation in arrays, that is true. But your proposed solution does the same thing

tbreloff commented:
I think we should try to get some outside opinions on this... could start a discussion on julia-users?

Evizero commented:
Sure. It won't change much for this PR though, as I am running out of time I can spend on it and I would like to not leave it half done

tbreloff commented:
Ok, well I think you should do what you want to do, and I'm going to go back to using what I have in StochasticOptimization for now because I can't use what you're proposing without refactoring other code.  Maybe we should just have 2 alternate APIs and move on with our lives...

Evizero commented:
I suppose it was just a matter of time before we encounter some decision that is not so easily settled. I am sure we will solve this down the line, given some time. Note, that I don't see my way as final, or this discussion as resolved. 

Also, even if we undo the `DataSubset` in the near or far future, I still think we made good progress in the right direction for a lot of things. And thanks for having this discussion in the first place.

tbreloff commented:
Yes agreed on all points... I think we made progress in design, even if we ended up on different sides. :)

I'm going to re-do the iteration inside StochasticOptimization as a sub-module that you have to explicitly import, so that a user could choose which approach to use.  And maybe someday we can unify the API more.  I'm going to use the ideas we discussed here to do:

``` julia
export
    DataIterator,
        ObsIterator,        # each iteration is an observation T
            EachObs,
            DataSubset,
            InfiniteObs,
        BatchIterator,      # each iteration is a ObsIterator{T}
            EachBatch,
            Batches,
            InfiniteBatches,
        BatchesIterator,    # each iteration is a BatchIterator{T}
            KFolds

abstract DataIterator{T} <: AbstractVector{T}
    abstract ObsIterator{T} <: DataIterator{T}
    abstract BatchIterator{T} <: DataIterator{T}
    abstract BatchesIterator{T} <: DataIterator{T}
```

Evizero commented:
so `infinite_obs` -> `InfiniteObs`, and `infinite_batches` -> `InfiniteBatches` ?

tbreloff commented:
FYI... subtyping from AbstractVector has the unexpected nicety of pretty printing:

``` julia
julia> using StochasticOptimization.Iteration

julia> train,test = split_obs(rand(10))
2-element StochasticOptimization.Iteration.Batches{Array{Float64,1},Float64,Array{UnitRange{Int64},1}}:
 [0.260174,0.772488,0.875043,0.605563,0.685558,0.923593,0.630983]
 [0.737278,0.963126,0.394369]                                    

julia> train
7-element StochasticOptimization.Iteration.SubsetObs{Array{Float64,1},Float64,UnitRange{Int64}}:
 0.260174
 0.772488
 0.875043
 0.605563
 0.685558
 0.923593
 0.630983
```

Evizero commented:
That does have some appeal. Certainly worth investigating. I wonder if there are any nuances to be aware of or contracts to obey. I think a long while ago I wanted to make something an abstract vector. Oh yea it was in ValueHistories [here in an experimental branch](https://github.com/JuliaML/ValueHistories.jl/blob/collections/src/vector_uv_history.jl#L1). I remember having issues because of that. But that may have just been me doing something stupid

Evizero commented:
It seems `setindex!` is not considered optional for abstract arrays: http://docs.julialang.org/en/release-0.5/manual/interfaces/#abstract-arrays. I suspect for some good reason.
Not sure how we would implement that

Evizero commented:
I am sold on AbstractArray by now. Main reason being that it allows using `StatsBase.sample`

``` julia
julia> sample(eachobs(X,y), 5)
5-element Array{Tuple{SubArray{Float64,1,Array{Float64,2},Tuple{Colon,Int64},true},String},1}:
 ([6.3,3.3,6.0,2.5],"virginica") 
 ([6.6,2.9,4.6,1.3],"versicolor")
 ([6.1,3.0,4.6,1.4],"versicolor")
 ([6.3,2.3,4.4,1.3],"versicolor")
 ([5.0,2.0,3.5,1.0],"versicolor")
```

Evizero commented:
`AbstractArray` as base-class has it's issues with infinite iterators though. I am experimenting with reproducible indexing for that.
